### PR TITLE
Add type hints for mmcv/engine/test.py

### PR DESCRIPTION
**Pull request by tcexeexe**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Modification

Add type hints in 
engine/test.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
> [CLA assistant check](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=2025) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=2025) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmcv?pullRequest=2025) it.

Hi @tcexeexe , thanks for your contributions. Please sign the CLA.

File: mmcv/engine/test.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 from mmcv.runner import get_dist_info
 
 
-def single_gpu_test(model, data_loader):
+def single_gpu_test(model, data_loader) -> list:
```
Comment (zhouzaida) [inline mmcv/engine/test.py:15]: Please add a type hint for `data_loader`.

File: mmcv/engine/test.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def single_gpu_test(model, data_loader):
     return results
 
 
-def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
+def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False) -> list:
```
Comment (zhouzaida) [inline mmcv/engine/test.py:44]: please add type hints for arguments

File: mmcv/engine/test.py [hunk 3]
```diff
@@ -88,7 +88,7 @@ def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
     return results
 
 
-def collect_results_cpu(result_part, size, tmpdir=None):
+def collect_results_cpu(result_part, size, tmpdir=None) -> list:
```
Comment (zhouzaida) [inline mmcv/engine/test.py:91]: same as above

File: mmcv/engine/test.py [hunk 4]
```diff
@@ -152,7 +152,7 @@ def collect_results_cpu(result_part, size, tmpdir=None):
         return ordered_results
 
 
-def collect_results_gpu(result_part, size):
+def collect_results_gpu(result_part, size) -> list:
```
Comment (zhouzaida) [inline mmcv/engine/test.py:155]: same as above

**Comment by tcexeexe** [comment]
> > [CLA assistant check](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=2025) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=2025) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmcv?pullRequest=2025) it.
> 
> Hi @tcexeexe , thanks for your contributions. Please sign the CLA.

@zhouzaida Hi , l have signed it.

File: mmcv/engine/test.py (hunk 1 above)
Comment (tcexeexe) [reply mmcv/engine/test.py]: > Please add a type hint for `data_loader`.

@zhouzaida Hello, l have add a type hint for all arguments

File: mmcv/engine/test.py (hunk 2 above)
Comment (tcexeexe) [reply mmcv/engine/test.py]: same as above

File: mmcv/engine/test.py (hunk 3 above)
Comment (tcexeexe) [reply mmcv/engine/test.py]: same as above

File: mmcv/engine/test.py (hunk 4 above)
Comment (tcexeexe) [reply mmcv/engine/test.py]: same as above

**Comment by zhouzaida** [comment]
Hi, the lint failed.

File: mmcv/engine/test.py [hunk 5]
```diff
@@ -4,15 +4,17 @@
 import shutil
 import tempfile
 import time
+from typing import Optional, Union
 
 import torch
 import torch.distributed as dist
+import torch.nn as nn
```
Comment (zhouzaida) [inline mmcv/engine/test.py:11]:
```suggestion
import torch.nn as nn
from torch.utils.data import DataLoader
```

File: mmcv/engine/test.py [hunk 6]
```diff
@@ -4,15 +4,17 @@
...
 
 import mmcv
 from mmcv.runner import get_dist_info
 
 
-def single_gpu_test(model, data_loader):
+def single_gpu_test(model: nn.Module, data_loader: nn.Dataloader) -> list:
```
Comment (zhouzaida) [inline mmcv/engine/test.py:17]:
```suggestion
def single_gpu_test(model: nn.Module, data_loader: DataLoader) -> list:
```

File: mmcv/engine/test.py [hunk 7]
```diff
@@ -41,7 +43,10 @@ def single_gpu_test(model, data_loader):
     return results
 
 
-def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
+def multi_gpu_test(model: nn.Module,
+                   data_loader: nn.Dataloader,
```
Comment (zhouzaida) [inline mmcv/engine/test.py:47]:
```suggestion
                   data_loader: DataLoader,
```

File: mmcv/engine/test.py [hunk 8]
```diff
@@ -41,7 +43,10 @@ def single_gpu_test(model, data_loader):
...
+                   tmpdir: Optional[str] = None,
+                   gpu_collect: Optional[bool] = False) -> list:
```
Comment (zhouzaida) [inline mmcv/engine/test.py:49]:
```suggestion
                   gpu_collect: bool = False) -> list:
```

**Comment by zhouzaida** [comment]
https://github.com/open-mmlab/mmcv/blob/569d588e155159fe8b9c5d9b08469a2814dd7e35/mmcv/engine/test.py#L55

```python
data_loader (Dataloader): Pytorch data loader. 
```

File: mmcv/engine/test.py (hunk 6 above)
Comment (zhouzaida) [reply mmcv/engine/test.py:17]: Need to fix the type of `data_loader` in docstring.

https://github.com/open-mmlab/mmcv/blob/569d588e155159fe8b9c5d9b08469a2814dd7e35/mmcv/engine/test.py#L55

```python
data_loader (Dataloader): Pytorch data loader. 
```

**Comment by tcexeexe** [comment]
Create a another PR.

**Pull request closed without merging**